File: mobile/plugins/qr-scanner/ios/Plugin/MgtQrScannerPlugin.swift
import Foundation
import AVFoundation
import Capacitor

/**
 * Native CCCD QR scanner. On iOS 16+ uses Apple VisionKit (DataScannerViewController,
 * the Camera-app engine); otherwise an AVFoundation AVCaptureMetadataOutput scanner.
 * Both return the decoded payload plus the locked frame as an image.
 */
@objc(MgtQrScannerPlugin)
public class MgtQrScannerPlugin: CAPPlugin {

    private weak var presentedVC: UIViewController?

    @objc func scan(_ call: CAPPluginCall) {
        let status = AVCaptureDevice.authorizationStatus(for: .video)
        switch status {
        case .authorized:
            presentScanner(call)
        case .notDetermined:
            AVCaptureDevice.requestAccess(for: .video) { granted in
                if granted {
                    self.presentScanner(call)
                } else {
                    call.reject("Camera permission denied", "denied")
                }
            }
        default:
            call.reject("Camera permission denied", "denied")
        }
    }

    @objc func cancel(_ call: CAPPluginCall) {
        dismissPresented()
        call.resolve()
    }

    private func presentScanner(_ call: CAPPluginCall) {
        let imageReturn = call.getString("imageReturn") ?? "file"
        let quality = call.getInt("imageQuality") ?? 85
        let autoZoom = call.getBool("autoZoom") ?? true
        let torchButton = call.getBool("torchButton") ?? true
        let strings = call.getObject("strings")
        let hint = (strings?["hint"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "Đưa mã QR trên CCCD vào khung"
        let cancelText = (strings?["cancel"] as? String).flatMap { $0.isEmpty ? nil : $0 } ?? "Hủy"

        let onResult: (QrScanOutcome) -> Void = { [weak self] outcome in
            guard let self = self else { return }
            self.dismissPresented()
            switch outcome {
            case .success(let raw, let jpeg, let width, let height, let engine):
                var ret = JSObject()
                ret["raw"] = raw
                ret["format"] = "QR_CODE"
                ret["engine"] = engine
                ret["imageWidth"] = width
                ret["imageHeight"] = height
                if let jpeg = jpeg {
                    if imageReturn == "base64" {
                        ret["imageBase64"] = jpeg.base64EncodedString()
                    } else if imageReturn == "file", let uri = self.writeTemp(jpeg) {
                        ret["savedUri"] = uri
                    }
                }
                call.resolve(ret)
            case .cancelled:
                call.reject("cancelled", "cancelled")
            case .failure(let message):
                call.reject(message, "error")
            }
        }

        DispatchQueue.main.async {
            let vc: UIViewController
            if #available(iOS 16.0, *), DataScannerScanner.isAvailable() {
                vc = DataScannerScanner(quality: quality, hint: hint, cancelText: cancelText, onResult: onResult)
            } else {
                vc = QRScannerViewController(quality: quality, autoZoom: autoZoom, torchButton: torchButton,
                                            hint: hint, cancelText: cancelText, onResult: onResult)
            }
            vc.modalPresentationStyle = .fullScreen
            self.presentedVC = vc
            self.bridge?.viewController?.present(vc, animated: true, completion: nil)
        }
    }

    private func dismissPresented() {
        DispatchQueue.main.async {
            self.presentedVC?.dismiss(animated: true, completion: nil)
            self.presentedVC = nil
        }
    }

    private func writeTemp(_ data: Data) -> String? {
        let name = "cccd-qr-\(Int(Date().timeIntervalSince1970 * 1000)).jpg"
        let url = FileManager.default.temporaryDirectory.appendingPathComponent(name)
        do {
            try data.write(to: url)
            return url.absoluteString
        } catch {
            return nil
        }
    }
}

File: mobile/plugins/qr-scanner/android/src/main/java/vn/motogiathinh/qrscanner/MgtQrScannerPlugin.kt
package vn.motogiathinh.qrscanner

import android.Manifest
import android.app.Activity
import android.content.Intent
import androidx.activity.result.ActivityResult
import com.getcapacitor.JSObject
import com.getcapacitor.PermissionState
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.ActivityCallback
import com.getcapacitor.annotation.CapacitorPlugin
import com.getcapacitor.annotation.Permission
import com.getcapacitor.annotation.PermissionCallback

@CapacitorPlugin(
    name = "MgtQrScanner",
    permissions = [Permission(strings = [Manifest.permission.CAMERA], alias = "camera")]
)
class MgtQrScannerPlugin : Plugin() {

    @PluginMethod
    fun scan(call: PluginCall) {
        if (getPermissionState("camera") != PermissionState.GRANTED) {
            requestPermissionForAlias("camera", call, "cameraPermsCallback")
            return
        }
        launchScanner(call)
    }

    @PermissionCallback
    fun cameraPermsCallback(call: PluginCall) {
        if (getPermissionState("camera") == PermissionState.GRANTED) {
            launchScanner(call)
        } else {
            call.reject("Camera permission denied", "denied")
        }
    }

    private fun launchScanner(call: PluginCall) {
        val intent = Intent(context, ScannerActivity::class.java).apply {
            putExtra("imageReturn", call.getString("imageReturn", "file"))
            putExtra("imageQuality", call.getInt("imageQuality", 85)!!)
            putExtra("autoZoom", call.getBoolean("autoZoom", true)!!)
            putExtra("torchButton", call.getBoolean("torchButton", true)!!)
            call.getObject("strings")?.let { s ->
                putExtra("hint", s.optString("hint", ""))
                putExtra("cancel", s.optString("cancel", ""))
            }
        }
        startActivityForResult(call, intent, "scanActivityResult")
    }

    @ActivityCallback
    fun scanActivityResult(call: PluginCall?, result: ActivityResult) {
        if (call == null) return
        val data = result.data
        if (result.resultCode == Activity.RESULT_OK && data != null && data.getStringExtra("raw") != null) {
            val ret = JSObject()
            ret.put("raw", data.getStringExtra("raw"))
            ret.put("format", "QR_CODE")
            ret.put("engine", "mlkit-camerax")
            data.getStringExtra("savedUri")?.let { ret.put("savedUri", it) }
            data.getStringExtra("imageBase64")?.let { ret.put("imageBase64", it) }
            if (data.hasExtra("imageWidth")) ret.put("imageWidth", data.getIntExtra("imageWidth", 0))
            if (data.hasExtra("imageHeight")) ret.put("imageHeight", data.getIntExtra("imageHeight", 0))
            call.resolve(ret)
        } else {
            call.reject("cancelled", "cancelled")
        }
    }

    @PluginMethod
    fun cancel(call: PluginCall) {
        // The scanner Activity owns its own lifecycle (user taps Hủy / back).
        call.resolve()
    }
}
